### PR TITLE
[clang][RISCV] Change default abi with f extension but without d extension

### PR DESCRIPTION
**Pull request by jacquesguan**
Now we have default abi lp64 for rv64if and ilp32 for rv32if, which is different with riscv-gnu-toolchain. In https://github.com/riscv-collab/riscv-gnu-toolchain/blob/8e9fb09a0c4b1e566492ee6f42e8c1fa5ef7e0c2/configure#L3385 when have f and not d, it prefers lp64f/ilp32f but no soft float. This patch tries to make their behaviors consistent.

### CODE REVIEW

**Review by wangpc-pp — APPROVED** [review]
I have similar patch before: https://reviews.llvm.org/D125947, so it LGTM as GCC's default behavior has changed. :-)
But please wait for others' opinions.

**Comment by kito-cheng** [comment]
short version: GCC isn't change. 
long version: GCC's configure script isn't change, it's configure script in riscv-gnu-toolchain

But I don't have strong opinion on this change since I believe user should explicitly specify that, otherwise it's really to screw up to select multi-lib or portability issue between different compilers.

**Comment by wangpc-pp** [comment]
> short version: GCC isn't change. long version: GCC's configure script isn't change, it's configure script in riscv-gnu-toolchain

So why is there a difference between GCC and riscv-gnu-toolchain? If we set `with_abi` to lp64f, what is the behavior?
> 
> But I don't have strong opinion on this change since I believe user should explicitly specify that, otherwise it's really to screw up to select multi-lib or portability issue between different compilers.

**Comment by jacquesguan** [comment]
soft ping... Any more advice?

**Review by MaskRay — APPROVED** [review]

**Comment by MaskRay** [comment]
> when only have f extension but no d extension

This can be simplified to "w/ f extension but w/o d extension"

File: clang/test/Driver/riscv-cpus.c [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Check target CPUs are correctly passed.
+·// Check target CPUs are correctly passed.
```
[the remark this answers is not in the document]
Comment (jacquesguan) [reply clang/test/Driver/riscv-cpus.c]: Fixed

**Comment by benshi001** [comment]
I think this is reasonable, even I myself made such a change months ago in Phabracator, but finally not approved.

**Review by benshi001 — APPROVED** [review]

**Comment by wangpc-pp** [comment]
As @kito-cheng has pointed out, we should fix multilib issue before landing this. It seems we don't generate multilib for ilp32f and lp64f in riscv-gnu-toolchain?

**Comment by asb** [comment]
This is a user-facing change that definitely needs to be acknowledged in the release notes (clang/docs/ReleaseNotes.rst)

I agree with you that it seems more intuitive that a -march=rv32imaf invocation should default to ilp32f just as -march=rv32imafd defaults to ilp32d. I slightly disagree that this patch aligns us to GCC, because as I understood Kito's comment the GCC behaviour is really depending on what default ABI the toolchain was configured with. For a toolchain that defaults to ilp32d, presumably -march=rv32imaf still ends up with ilp32?

That said, with the GCC cross-compilation model being so different to clang I don't think we necessarily need to match precisely, so I'm not opposed to a more intuitive default.

**Comment by jacquesguan** [comment]
> I think the conclusion from the LLVM sync-up call was that everyone happy to move in this direction, so please add the release note and we can do a final review. Thanks!

Done, added release note.

**Comment by jacquesguan** [comment]
> > > I think the conclusion from the LLVM sync-up call was that everyone happy to move in this direction, so please add the release note and we can do a final review. Thanks!
> 
> > 
> 
> > Done, added release note.
> 
> 
> 
> Thanks! Sorry I wasn't specific about this, but we need a Clang release note as well.

Done, added Clang release note too.

**Review by asb — APPROVED** [review]
LGTM

**Pull request merged**